erts/emulator/beam/erl_bif_lists.c

```diff
@@ -146,93 +157,538 @@ BIF_RETTYPE append_2(BIF_ALIST_2)
+                BIF_TRAP2(&subtract_trap_export, p,
+                          SUBTRACT_SPECIAL_VALUE, a_b_state);
+            }
+            count = signed_val(res);
+            /* Returning true means length is done, result in s.len.count */
+            if (count < 0) {
+                p->flags &= ~F_DISABLE_GC;
+                BIF_ERROR(p, BADARG);
+            }
+            if (count == 0) {
+                p->flags &= ~F_DISABLE_GC;
+                BIF_RET(NIL);
+            }
+            context->len_a = count;
+            context->fsm_state = SUBTRACT_STEP_LEN_B;
+            erts_trappable_list_length_setup(orig_B, context);
+            /* FALL THROUGH */
+        }
+
+        case SUBTRACT_STEP_LEN_B: {
+            Eterm res = erts_trappable_list_length(&context->s.len);
+            Sint count;
+            if (res == am_false) {
+                BIF_TRAP2(&subtract_trap_export, p,
+                          SUBTRACT_SPECIAL_VALUE, a_b_state);
+            }
+            count = signed_val(res);;
+            /* Returning true means length is done, result in s.len.count */
+            if (count < 0) {
+                p->flags &= ~F_DISABLE_GC;
+                BIF_ERROR(p, BADARG);
+            }
+            if (count == 0) {
+                p->flags &= ~F_DISABLE_GC;
+                BIF_RET(orig_A);
+            }
+            context->len_b = count;
+            context->fsm_state = SUBTRACT_STEP_COPY_B;
+            subtract_enter_state_copy_B(orig_B, context);
+            /* FALL THROUGH */
+        }
+
+        case SUBTRACT_STEP_COPY_B: {
+            if (!erts_trappable_copy_list_to_array(&context->s.copy_b)) {
+                BIF_TRAP2(&subtract_trap_export, p,
+                          SUBTRACT_SPECIAL_VALUE, a_b_state);
+            }
+            context->fsm_state = SUBTRACT_STEP_SCAN;
+            subtract_enter_state_scan(orig_A, context);
+            /* FALL THROUGH */
+        }
+
+        case SUBTRACT_STEP_SCAN: {
+            int res = subtract_trappable_scan(p, orig_A, context);
+            if (!res) {
+                BIF_TRAP2(&subtract_trap_export, p,
+                          SUBTRACT_SPECIAL_VALUE, a_b_state);
+            }
+            if (context->s.scan.result_size == 0) { /* All deleted? Return a [] */
+                p->flags &= ~F_DISABLE_GC;
+                BIF_RET(NIL);
+            } else if (context->s.scan.result_size == context->len_a) {
+                /* None deleted? Return the original A */
+                p->flags &= ~F_DISABLE_GC;
+                BIF_RET(orig_A);
+            }
+            context->fsm_state = SUBTRACT_STEP_BUILD_RESULT;
+            erts_trappable_copy_array_to_list_setup(p, context->result_array,
+                                                    context->s.scan.result_size,
+                                                    &context->s.build_res);
+            /* FALL THROUGH */
+        }
+
+        case SUBTRACT_STEP_BUILD_RESULT: {
```

Comment:
I think you could skip this step if you build the list destructively during SUBTRACT_STEP_SCAN instead. You've already disabled the GC, so it should be safe as long as you never leak the unfinished result.

Reply:
The problem is I do not know the size of the result beforehand, i.e. i'll have to assume heap is large enough and then allocate after the result is finished. Or if the heap is not large enough I'm in trouble.

Reply:
It's safe to `HAlloc` each element as you go along. This could result in a lot of heap fragments being created, but since you already know the maximum possible length of the result you can use `HAllocX` to reserve (but not allocate) all you need up-front.

Reply:
Ok let's say I really want to do this and will see if it can be done before I am gone on Sunday (traveling to work at the customer). The idea of not copying and building result in place is a really bright one.